added image processing extras

likangning93 opened the pull request:
In anticipation of ambient occlusion and other future image processing stages, I added extras for use with jimp (https://www.npmjs.com/package/jimp) in the form of a gltf-level scratch `Jimp` image and `Jimp` copies of all the existing images, stored as `jimpImage` in `extras._pipeline.`

Since this won't be used for every pipeline configuration, I tried to make it as optional as possible.

I also added relevant specs.

lilleyse commented:
I'm a bit torn, because while it's faster to make the image processing an optional step, it makes the code a bit more confusing. It would be kind of nice if image processing was completely hidden from `gltfPipeline.js`, and instead it always decoded in `loadGltfUris` and encodes in `writeSource` (but only if the data was changed, this can be marked by a flag in the image's `extras` that is set by the AO stage/atlas stage etc).

An alternative is to continue to allow `readGltf` and `loadGltfUris` to take the `imageProcess` flag, but not have an `encodeImages` stage in `gltfPipeline`, instead always encode in `writeSource` using the flag idea above.

Let me know what you think, or if I'm overlooking something.

pjcozzi commented:
Can't this be handled with each stage declaring its dependencies?  See #75.

likangning93 commented:
I really like option 2 for re-encoding. I also toyed with the idea of
changing 'loadGltfUris' to take an options parameter instead of 'basePath',
which might let us hide the image processing more. I'll give it a try.

On Thursday, June 16, 2016, Sean Lilley <notifications@github.com> wrote:

> I'm a bit torn, because while it's faster to make the image processing an
> optional step, it makes the code a bit more confusing. It would be kind of
> nice if image processing was completely hidden from gltfPipeline.js, and
> instead it always decoded in loadGltfUris and encodes in writeSource (but
> only if the data was changed, this can be marked by a flag in the image's
> extras that is set by the AO stage/atlas stage etc).
>
> An alternative is to continue to allow readGltf and loadGltfUris to take
> the imageProcess flag, but not have an encodeImages stage in gltfPipeline,
> instead always encode in writeSource using the flag idea above.
>
> Let me know what you think, or if I'm overlooking something.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/AnalyticalGraphicsInc/gltf-pipeline/pull/91#issuecomment-226628030>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AFJTN9BRgESGRtXNUIqnPzTutqOSCaWkks5qMckYgaJpZM4I3xQl>
> .
>

likangning93 commented:
Updated! As discussed offline, we decided to keep the `encodeImages` stage where it is for now. It seems like `writeSource` (which handles moving extras back into the json) is currently tied to `writeGltf` and not to `writeBinaryGltf`, and we may want to move it to come before the actual file writing options. When that happens we will move `encodeImages` over to that. `encodeImages` also now checks for an `imageChanged` flag before encoding.

I also updated `loadGltfUris` to take in an options parameter instead of just `basePath` and updated all relevant specs accordingly.

likangning93 commented:
merged in master.

lilleyse commented on lib/encodeImages.js:
```diff
@@ -0,0 +1,58 @@
+var Jimp = require('jimp');
+
+module.exports = encodeImages;
+var async = require('async');
```
Fix the ordering here.

lilleyse commented on lib/encodeImages.js:
```diff
@@ -0,0 +1,58 @@
+module.exports = encodeImages;
+var async = require('async');
+
+function encodeImages(gltfWithExtras, callback) {
```
Change `gltfWithExtras` to `gltf` to follow the other stages.

lilleyse commented on lib/encodeImages.js:
```diff
@@ -0,0 +1,58 @@
+        }
+        else if (callback) {
+            callback(gltfWithExtras);
+        }
```
Right now the callback to `encodeImage` expects a gltf object, not an error. Here you can just simplify to

```javascript
if (err) {
    throw err;
}
callback(gltf)
```

lilleyse commented on lib/loadGltfUris.js:
```diff
@@ -8,18 +8,37 @@ var DeveloperError = Cesium.DeveloperError;
+
     async.each(['buffers', 'images', 'shaders'], function(name, asyncCallback) {
-        loadURI(gltf, basePath, name, asyncCallback);
+        loadURI(gltf, basePath, name, asyncCallback, imageProcess);
```
Usually the callback should be the last argument.

lilleyse commented on lib/readGltf.js:
```diff
@@ -10,14 +10,13 @@ var DeveloperError = Cesium.DeveloperError;
 module.exports = readGltf;
 
-function readGltf(gltfPath, callback) {
+function readGltf(gltfPath, options, callback, imageProcess) {
```
You don't use `imageProcess` anymore right?

likangning93 replied:
It's still needed to signal that a `jimp` needs to get added.

lilleyse replied:
I mean, isn't it now `options.imageProcess`?

likangning93 replied:
I kind of unpacked the options further up, should I just be doing `options.xxx`?

likangning93 replied:
Ah. I was looking at the wrong file.

lilleyse commented on lib/readGltf.js:
```diff
@@ -35,11 +34,11 @@ function readGltf(gltfPath, callback) {
             }
             callback(gltf);
-        });
+        }, imageProcess);
```
Same comment here.

lilleyse commented on lib/readGltf.js:
```diff
@@ -10,14 +10,13 @@ var DeveloperError = Cesium.DeveloperError;
     }
     
     var fileExtension = path.extname(gltfPath);
-    var filePath = path.dirname(gltfPath);
```
I know this change is now everywhere, but I think `options.basePath` should just be an optional argument for `readGltf`. Almost all of the time the `basePath` can be computed like this.

lilleyse commented on specs/lib/encodeImagesSpec.js:
```diff
@@ -0,0 +1,85 @@
+var Jimp = require('jimp');
+
+describe('encodeImages', function() {
+    var imageData;
```
This var isn't used.

lilleyse commented on specs/lib/encodeImagesSpec.js:
```diff
@@ -0,0 +1,85 @@
+                expect(bufferEqual(pipelineExtras0002.source, imageBuffer)).toBe(true);
+                done();
+            });
+        }, true);
```
Remove this last parameter.

lilleyse commented on specs/lib/encodeImagesSpec.js:
```diff
@@ -0,0 +1,85 @@
+                    });
+                });
+            });
+        }, true);
```
And this.

lilleyse commented on specs/lib/gltfPipelineSpec.js:
```diff
@@ -91,7 +101,10 @@ describe('gltfPipeline', function() {
-        var options = { 'binary' : true };
+        var options = {
+            basePath: basePath,
+            'binary' : true
```
Remove quotation marks.

likangning93 commented:
Updated!

likangning93 commented:
@pjcozzi integrating image processing into `runStages` might be a little messy since the encoding and decoding have to happen at the beginning and end of the pipeline and it seems like `runStages` handles the stages that are more synchronous.

I imagine we could add something before runStages that checks if the pipeline needs image processing extras, but that could get messy if we have to add anything else that wraps the pipeline like this.

lilleyse commented on specs/lib/cacheOptimizeIndicesSpec.js:
```diff
@@ -5,6 +5,7 @@ var readGltf = require('../../lib/readGltf');
 var readAccessor = require('../../lib/readAccessor');
 var writeAccessor = require('../../lib/writeAccessor');
 
+var basePath = './specs/data/boxTexturedUnoptimized/';
```
There are a few lines like this that aren't used. Otherwise this PR looks ready to me.

likangning93 commented:
ok. Updated!

lasalvavida commented:
> @pjcozzi integrating image processing into runStages might be a little messy since the encoding and decoding have to happen at the beginning and end of the pipeline and it seems like runStages handles the stages that are more synchronous.

We should adopt a promise style architecture for both synchronous and asynchronous stages; it would allow all stages to be run in the same fashion.

pjcozzi commented:
> We should adopt a promise style architecture for both synchronous and asynchronous stages; it would allow all stages to be run in the same fashion.

Yes, please add it to #1.

pjcozzi commented on lib/gltfPipeline.js:
```diff
@@ -54,11 +55,16 @@ function processJSONWithExtras(gltfWithExtras, options, callback) {
-    callback(gltfWithExtras);
+    if (options.imageProcess) {
+        encodeImages(gltfWithExtras, callback);
+    }
```
`} else {`

pjcozzi commented on package.json:
```diff
@@ -38,6 +38,7 @@
     "data-uri-to-buffer": "0.0.4",
     "datauri": "^1.0.0-alpha.5",
     "image-size": "^0.5.0",
+    "jimp": "^0.2.24",
```
Add this to the third-party section of [LICENSE.md](https://github.com/AnalyticalGraphicsInc/gltf-pipeline/blob/master/LICENSE.md).  In general, can we please update that section to match the libraries here in `dependencies`?
